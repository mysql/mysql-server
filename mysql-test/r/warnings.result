set @@session.sql_mode='';
drop table if exists t1, t2;
SET SQL_WARNINGS=1;
create table t1 (a int);
create table t1 (a int);
ERROR 42S01: Table 't1' already exists
show count(*) errors;
@@session.error_count
1
show errors;
Level	Code	Message
Error	1050	Table 't1' already exists
show warnings;
Level	Code	Message
Error	1050	Table 't1' already exists
create table t2(a int) default charset qwerty;
ERROR 42000: Unknown character set: 'qwerty'
show count(*) errors;
@@session.error_count
1
show errors;
Level	Code	Message
Error	1115	Unknown character set: 'qwerty'
create table t (i);
ERROR 42000: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ')' at line 1
show count(*) errors;
@@session.error_count
1
show errors;
Level	Code	Message
Error	1064	You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ')' at line 1
insert into t1 values (1);
insert ignore into t1 values ("hej");
Warnings:
Warning	1366	Incorrect integer value: 'hej' for column 'a' at row 1
insert ignore into t1 values ("hej"),("d책");
Warnings:
Warning	1366	Incorrect integer value: 'hej' for column 'a' at row 1
Warning	1366	Incorrect integer value: 'd책' for column 'a' at row 2
set SQL_WARNINGS=1;
insert ignore into t1 values ("hej");
Warnings:
Warning	1366	Incorrect integer value: 'hej' for column 'a' at row 1
insert ignore into t1 values ("hej"),("d책");
Warnings:
Warning	1366	Incorrect integer value: 'hej' for column 'a' at row 1
Warning	1366	Incorrect integer value: 'd책' for column 'a' at row 2
drop table t1;
set SQL_WARNINGS=0;
drop temporary table if exists not_exists;
Warnings:
Note	1051	Unknown table 'test.not_exists'
drop table if exists not_exists_table;
Warnings:
Note	1051	Unknown table 'test.not_exists_table'
show warnings limit 1;
Level	Code	Message
Note	1051	Unknown table 'test.not_exists_table'
drop database if exists not_exists_db;
Warnings:
Note	1008	Can't drop database 'not_exists_db'; database doesn't exist
show count(*) warnings;
@@session.warning_count
1
create table t1(id int);
create table if not exists t1(id int);
Warnings:
Note	1050	Table 't1' already exists
select @@warning_count;
@@warning_count
1
drop table t1;
create table t1(a tinyint, b int not null, c date, d char(5));
load data infile '../../std_data/warnings_loaddata.dat' into table t1 fields terminated by ',';
Warnings:
Warning	1263	Column set to default value; NULL supplied to NOT NULL column 'b' at row 2
Warning	1265	Data truncated for column 'd' at row 3
Warning	1265	Data truncated for column 'c' at row 4
Warning	1261	Row 5 doesn't contain data for all columns
Warning	1265	Data truncated for column 'b' at row 6
Warning	1262	Row 7 was truncated; it contained more data than there were input columns
Warning	1264	Out of range value for column 'a' at row 8
select @@warning_count;
@@warning_count
7
drop table t1;
create table t1(a tinyint NOT NULL, b tinyint unsigned, c char(5));
insert ignore into t1 values(NULL,100,'mysql'),(10,-1,'mysql ab'),(500,256,'open source'),(20,NULL,'test');
Warnings:
Warning	1048	Column 'a' cannot be null
Warning	1264	Out of range value for column 'b' at row 2
Warning	1265	Data truncated for column 'c' at row 2
Warning	1264	Out of range value for column 'a' at row 3
Warning	1264	Out of range value for column 'b' at row 3
Warning	1265	Data truncated for column 'c' at row 3
SET sql_mode = 'NO_ENGINE_SUBSTITUTION';
alter table t1 modify c char(4);
Warnings:
Warning	1265	Data truncated for column 'c' at row 1
Warning	1265	Data truncated for column 'c' at row 2
SET sql_mode = default;
alter table t1 add d char(2);
update ignore t1 set a=NULL where a=10;
Warnings:
Warning	1048	Column 'a' cannot be null
update ignore t1 set c='mysql ab' where c='test';
Warnings:
Warning	1265	Data truncated for column 'c' at row 4
update ignore t1 set d=c;
Warnings:
Warning	1265	Data truncated for column 'd' at row 1
Warning	1265	Data truncated for column 'd' at row 2
Warning	1265	Data truncated for column 'd' at row 3
Warning	1265	Data truncated for column 'd' at row 4
create table t2(a tinyint NOT NULL, b char(3));
insert ignore into t2 select b,c from t1;
Warnings:
Warning	1265	Data truncated for column 'b' at row 1
Warning	1265	Data truncated for column 'b' at row 2
Warning	1265	Data truncated for column 'b' at row 3
Warning	1048	Column 'a' cannot be null
Warning	1265	Data truncated for column 'b' at row 4
insert ignore into t2(b) values('mysqlab');
Warnings:
Warning	1265	Data truncated for column 'b' at row 1
Warning	1364	Field 'a' doesn't have a default value
set sql_warnings=1;
insert ignore into t2(b) values('mysqlab');
Warnings:
Warning	1265	Data truncated for column 'b' at row 1
Warning	1364	Field 'a' doesn't have a default value
set sql_warnings=0;
drop table t1, t2;
create table t1(a char(10));
alter table t1 add b char;
set max_error_count=10;
update ignore t1 set b=a;
Warnings:
Warning	1265	Data truncated for column 'b' at row 1
Warning	1265	Data truncated for column 'b' at row 2
Warning	1265	Data truncated for column 'b' at row 3
Warning	1265	Data truncated for column 'b' at row 4
Warning	1265	Data truncated for column 'b' at row 5
Warning	1265	Data truncated for column 'b' at row 6
Warning	1265	Data truncated for column 'b' at row 7
Warning	1265	Data truncated for column 'b' at row 8
Warning	1265	Data truncated for column 'b' at row 9
Warning	1265	Data truncated for column 'b' at row 10
select @@warning_count;
@@warning_count
50
set max_error_count=0;
show variables like 'max_error_count';
Variable_name	Value
max_error_count	0
update ignore t1 set b='hi';
select @@warning_count;
@@warning_count
50
show warnings;
Level	Code	Message
set max_error_count=65535;
show variables like 'max_error_count';
Variable_name	Value
max_error_count	65535
set max_error_count=10;
show variables like 'max_error_count';
Variable_name	Value
max_error_count	10
drop table t1;
create table t1 (a int);
insert into t1 (a) values (1), (2), (3), (4), (5), (6), (7), (8), (9), (10);
update ignore t1 set a='abc';
Warnings:
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 1
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 2
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 3
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 4
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 5
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 6
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 7
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 8
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 9
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 10
show warnings limit 2, 1;
Level	Code	Message
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 3
show warnings limit 0, 10;
Level	Code	Message
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 1
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 2
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 3
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 4
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 5
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 6
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 7
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 8
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 9
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 10
show warnings limit 9, 1;
Level	Code	Message
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 10
show warnings limit 10, 1;
Level	Code	Message
show warnings limit 9, 2;
Level	Code	Message
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 10
show warnings limit 0, 0;
Level	Code	Message
show warnings limit 1;
Level	Code	Message
Warning	1366	Incorrect integer value: 'abc' for column 'a' at row 1
show warnings limit 0;
Level	Code	Message
show warnings limit 1, 0;
Level	Code	Message
select * from t1 limit 0;
a
select * from t1 limit 1, 0;
a
select * from t1 limit 0, 0;
a
drop table t1;
End of 4.1 tests
CREATE TABLE t1( f1 CHAR(20) );
CREATE TABLE t2( f1 CHAR(20), f2 CHAR(25) );
CREATE TABLE t3( f1 CHAR(20), f2 CHAR(25), f3 DATE );
INSERT INTO t1 VALUES ( 'a`' );
INSERT INTO t2 VALUES ( 'a`', 'a`' );
INSERT INTO t3 VALUES ( 'a`', 'a`', '1000-01-1' );
DROP PROCEDURE IF EXISTS sp1;
Warnings:
Note	1305	PROCEDURE test.sp1 does not exist
DROP PROCEDURE IF EXISTS sp2;
Warnings:
Note	1305	PROCEDURE test.sp2 does not exist
DROP PROCEDURE IF EXISTS sp3;
Warnings:
Note	1305	PROCEDURE test.sp3 does not exist
SET sql_mode = 'NO_ENGINE_SUBSTITUTION';
CREATE PROCEDURE sp1()
BEGIN
DECLARE x NUMERIC ZEROFILL;
SELECT f1 INTO x FROM t1 LIMIT 1;
END//
Warnings:
Warning	1681	The ZEROFILL attribute is deprecated and will be removed in a future release. Use the LPAD function to zero-pad numbers, or store the formatted numbers in a CHAR column.
CREATE PROCEDURE sp2()
BEGIN
DECLARE x NUMERIC ZEROFILL;
SELECT f1 INTO x FROM t2 LIMIT 1;
END//
Warnings:
Warning	1681	The ZEROFILL attribute is deprecated and will be removed in a future release. Use the LPAD function to zero-pad numbers, or store the formatted numbers in a CHAR column.
CREATE PROCEDURE sp3()
BEGIN
DECLARE x NUMERIC ZEROFILL;
SELECT f1 INTO x FROM t3 LIMIT 1;
END//
Warnings:
Warning	1681	The ZEROFILL attribute is deprecated and will be removed in a future release. Use the LPAD function to zero-pad numbers, or store the formatted numbers in a CHAR column.
CALL sp1();
Warnings:
Warning	1366	Incorrect decimal value: 'a`' for column 'x' at row 1
CALL sp2();
Warnings:
Warning	1366	Incorrect decimal value: 'a`' for column 'x' at row 1
CALL sp3();
Warnings:
Warning	1366	Incorrect decimal value: 'a`' for column 'x' at row 1
DROP PROCEDURE IF EXISTS sp1;
CREATE PROCEDURE sp1()
BEGIN
declare x numeric unsigned zerofill;
SELECT f1 into x from t2 limit 1;
END//
Warnings:
Warning	1681	The ZEROFILL attribute is deprecated and will be removed in a future release. Use the LPAD function to zero-pad numbers, or store the formatted numbers in a CHAR column.
Warning	1681	UNSIGNED for decimal and floating point data types is deprecated and support for it will be removed in a future release.
CALL sp1();
Warnings:
Warning	1366	Incorrect decimal value: 'a`' for column 'x' at row 1
DROP TABLE t1;
DROP TABLE t2;
DROP TABLE t3;
DROP PROCEDURE sp1;
DROP PROCEDURE sp2;
DROP PROCEDURE sp3;
SET sql_mode = default;
create table t1 (c_char char(255), c_varchar varchar(255), c_tinytext tinytext);
create table t2 (c_tinyblob tinyblob);
set @c = repeat(' ', 256);
set @q = repeat('q', 256);
set sql_mode = '';
insert into t1 values(@c, @c, @c);
Warnings:
Note	1265	Data truncated for column 'c_varchar' at row 1
Note	1265	Data truncated for column 'c_tinytext' at row 1
insert into t2 values(@c);
Warnings:
Warning	1265	Data truncated for column 'c_tinyblob' at row 1
insert into t1 values(@q, @q, @q);
Warnings:
Warning	1265	Data truncated for column 'c_char' at row 1
Warning	1265	Data truncated for column 'c_varchar' at row 1
Warning	1265	Data truncated for column 'c_tinytext' at row 1
insert into t2 values(@q);
Warnings:
Warning	1265	Data truncated for column 'c_tinyblob' at row 1
set sql_mode = 'traditional';
insert into t1 values(@c, @c, @c);
Warnings:
Note	1265	Data truncated for column 'c_varchar' at row 1
Note	1265	Data truncated for column 'c_tinytext' at row 1
insert into t2 values(@c);
ERROR 22001: Data too long for column 'c_tinyblob' at row 1
insert into t1 values(@q, NULL, NULL);
ERROR 22001: Data too long for column 'c_char' at row 1
insert into t1 values(NULL, @q, NULL);
ERROR 22001: Data too long for column 'c_varchar' at row 1
insert into t1 values(NULL, NULL, @q);
ERROR 22001: Data too long for column 'c_tinytext' at row 1
insert into t2 values(@q);
ERROR 22001: Data too long for column 'c_tinyblob' at row 1
drop table t1, t2;
DROP TABLE t1;
ERROR 42S02: Unknown table 'test.t1'
SHOW ERRORS;
Level	Code	Message
Error	1051	Unknown table 'test.t1'
End of 5.0 tests

-- Bug#55847

DROP TABLE IF EXISTS t1;
DROP FUNCTION IF EXISTS f1;
CREATE TABLE t1(a INT UNIQUE);
CREATE FUNCTION f1(x INT) RETURNS INT
BEGIN
INSERT INTO t1 VALUES(x);
INSERT INTO t1 VALUES(x);
RETURN x;
END|

SHOW TABLES WHERE f1(11) = 11;
ERROR 23000: Duplicate entry '11' for key 'a'

SHOW WARNINGS;
Level	Code	Message
Error	1062	Duplicate entry '11' for key 'a'
Error	1028	Sort aborted: Duplicate entry '11' for key 'a'

DROP TABLE t1;
DROP FUNCTION f1;
