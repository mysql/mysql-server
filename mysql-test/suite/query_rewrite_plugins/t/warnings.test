--echo #
--echo # Test of rules where the parser gives warnings about the originial
--echo # query, pattern or replacement.
--echo #

call mtr.add_suppression("Plugin rewriter reported: 'Some rules failed to load.'");

--source suite/query_rewrite_plugins/include/have_plugin_rewriter.inc
--source suite/query_rewrite_plugins/include/install_rewriter.inc

# Still many unsupported statements
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'LOAD DATA INFILE "tab1.data" INTO TABLE test.t1',
         'LOAD DATA INFILE "tab2.data" INTO TABLE test.t1' ),
       ( 'DO CONCAT("a", "b")', 'DO CONCAT("c", "d")' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();
SELECT * FROM query_rewrite.rewrite_rules;

INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SHOW /* I rewrote it */ WARNINGS' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();

--echo # We want the note about the query being rewritten here ...
SHOW WARNINGS;
SHOW ERRORS;
GET DIAGNOSTICS CONDITION 1 @p1 = RETURNED_SQLSTATE, @p2 = MESSAGE_TEXT;
SELECT @p1;
SELECT @p2;

--echo # ... but not here, since we failed parsing after seeing SHOW WARNINGS.
--error ER_PARSE_ERROR
SHOW WARNINGS and give me a parse error, please;

DELETE FROM query_rewrite.rewrite_rules;

--echo # Test of warnings when we rewrite the diagnostics statement.  echo #
--echo # We disable mtr's SHOW WARNINGS here, as it would just be confusing
--echo # when it gets rewritten without us seeing it.
--disable_warnings
INSERT INTO query_rewrite.rewrite_rules ( pattern, replacement )
VALUES ( 'SHOW WARNINGS', 'SELECT "used to be SHOW WARNINGS"' ),
       ( 'SELECT 1', 'SHOW WARNINGS' );

--error ER_SIGNAL_EXCEPTION
CALL query_rewrite.flush_rewrite_rules();

SELECT 1;

SHOW WARNINGS;

--enable_warnings
--source suite/query_rewrite_plugins/include/uninstall_rewriter.inc
