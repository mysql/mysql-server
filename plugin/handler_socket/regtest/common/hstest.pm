
# vim:sw=2:ai

package hstest;

use DBI;
use Net::HandlerSocket;

our %conf = ();

sub get_conf_env {
  my ($key, $defval) = @_;
  return $ENV{$key} || $defval;
}

sub init_conf {
  $conf{host} = get_conf_env("MYHOST", "localhost");
  $conf{myport} = get_conf_env("MYPORT", 3306);
  $conf{dbname} = get_conf_env("MYDBNAME", "hstestdb");
  $conf{ssps} = get_conf_env("MYSSPS");
  $conf{user} = get_conf_env("MYSQLUSER", "root");
  $conf{pass} = get_conf_env("MYSQLPASS", "");
  $conf{hsport} = get_conf_env("HSPORT", 9998);
  $conf{hspass} = get_conf_env("HSPASS", undef);
}

sub get_dbi_connection {
  my ($dbname, $host, $myport, $ssps, $user, $pass)
    = ($conf{dbname}, $conf{host}, $conf{myport}, $conf{ssps},
      $conf{user}, $conf{pass});
  my $mycnf = "binary_my.cnf";
  my $dsn = "DBI:mysql:database=;host=$host;port=$myport"
    . ";mysql_server_prepare=$ssps"
    . ";mysql_read_default_group=perl"
    . ";mysql_read_default_file=../common/$mycnf";
  my $dbh = DBI->connect($dsn, $user, $pass, { RaiseError => 1 });
  return $dbh;
}

sub init_testdb {
  my $charset = $_[0] || "binary";
  my $dbh = get_dbi_connection();
  my $dbname = $conf{dbname};
  $dbh->do("drop database if exists $dbname");
  $dbh->do("create database $dbname default character set $charset");
  $dbh->do("use $dbname");
  return $dbh;
}

sub get_hs_connection {
  my ($host, $port) = @_;
  $host ||= $conf{host};
  $port ||= $conf{hsport};
  my $hsargs = { 'host' => $host, 'port' => $port };
  my $conn = new Net::HandlerSocket($hsargs);
  if (defined($conn) && defined($conf{hspass})) {
    $conn->auth($conf{hspass});
  }
  return $conn;
}


init_conf();

1;

