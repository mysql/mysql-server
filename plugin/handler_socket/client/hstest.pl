#!/usr/bin/perl

# vim:sw=8:ai:ts=8

use strict;
use warnings;

use DBI;
use Net::HandlerSocket;

my %conf = ();
for my $i (@ARGV) {
	my ($k, $v) = split(/=/, $i);
	$conf{$k} = $v;
}

my $verbose = get_conf("verbose", 0);
my $actions_str = get_conf("actions", "hsread");
my $tablesize = get_conf("tablesize", 10000);
my $db = get_conf("db", "hstest");
my $table = get_conf("table", "hstest_table1");
my $engine = get_conf("engine", "innodb");
my $host = get_conf("host", "localhost");
my $mysqlport = get_conf("mysqlport", 3306);
my $mysqluser = get_conf("mysqluser", "root");
my $mysqlpass = get_conf("mysqlpass", "");
my $hsport = get_conf("hsport", 9999);
my $loop = get_conf("loop", 10000);
my $op = get_conf("op", "=");
my $ssps = get_conf("ssps", 0);
my $num_moreflds = get_conf("moreflds", 0);
my $moreflds_prefix = get_conf("moreflds_prefix", "column0123456789_");
my $keytype = get_conf("keytype", "varchar(32)");
my $file = get_conf("file", undef);

my $dsn = "DBI:mysql:database=;host=$host;port=$mysqlport"
	. ";mysql_server_prepare=$ssps";
my $dbh = DBI->connect($dsn, $mysqluser, $mysqlpass, { RaiseError => 1 });
my $hsargs = { 'host' => $host, 'port' => $hsport };
my $cli = new Net::HandlerSocket($hsargs);

my @actions = split(/,/, $actions_str);
for my $action (@actions) {
	if ($action eq "table") {
		print("TABLE $db.$table\n");
		$dbh->do("drop database if exists $db");
		$dbh->do("create database $db");
		$dbh->do("use $db");
		my $moreflds = get_createtbl_moreflds_str();
		$dbh->do(
			"create table $table (" .
			"k $keytype primary key" .
			",v varchar(32) not null" .
			$moreflds .
			") character set utf8 collate utf8_bin " .
			"engine = $engine");
	} elsif ($action eq "insert") {
		print("INSERT $db.$table tablesize=$tablesize\n");
		$dbh->do("use $db");
		my $moreflds = get_insert_moreflds_str();
		for (my $i = 0; $i < $tablesize; $i += 100) {
			my $qstr = "insert into $db.$table values";
			for (my $j = 0; $j < 100; ++$j) {
				if ($j != 0) {
					$qstr .= ",";
				}
				my $k = "" . ($i + $j);
				my $v = "v" . int(rand(1000)) . ($i + $j);
				$qstr .= "('$k', '$v'";
				for (my $j = 0; $j < $num_moreflds; ++$j) {
					$qstr .= ",'$j'";
				}
				$qstr .= ")";
			}
			$dbh->do($qstr);
			print "$i/$tablesize\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "read") {
		print("READ $db.$table op=$op loop=$loop\n");
		$dbh->do("use $db");
		my $moreflds = get_select_moreflds_str();
		my $sth = $dbh->prepare(
			"select k,v$moreflds from $db.$table where k = ?");
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			# print "k=$k\n";
			$sth->execute($k);
			if ($verbose >= 10) {
				print "RET:";
				while (my $ref = $sth->fetchrow_arrayref()) {
					my $rk = $ref->[0];
					my $rv = $ref->[1];
					print " $rk $rv";
				}
				print "\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsinsert") {
		print("HSINSERT $db.$table tablesize=$tablesize\n");
		$cli->open_index(1, $db, $table, '', 'k,v');
		for (my $i = 0; $i < $tablesize; ++$i) {
			my $k = "" . $i;
			my $v = "v" . int(rand(1000)) . $i;
			my $r = $cli->execute_insert(1, [ $k, $v ]);
			if ($r->[0] != 0) {
				die;
			}
			print "$i/$tablesize\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsread") {
		print("HSREAD $db.$table op=$op loop=$loop\n");
		my $moreflds = get_select_moreflds_str();
		$cli->open_index(1, $db, $table, '', "k,v$moreflds");
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			# print "k=$k\n";
			my $r = $cli->execute_find(1, $op, [ $k ], 1, 0);
			if ($verbose >= 10) {
				my $len = scalar(@{$r});
				print "LEN=$len";
				for my $e (@{$r}) {
					print " [$e]";
				}
				print "\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsupdate") {
		my $vbase = "v" . int(rand(1000));
		print("HSUPDATE $db.$table op=$op loop=$loop vbase=$vbase\n");
		$cli->open_index(1, $db, $table, '', 'v');
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			my $v = $vbase . $i;
			print "k=$k v=$v\n";
			my $r = $cli->execute_update(1, $op, [ $k ], 1, 0,
				[ $v ]);
			if ($verbose >= 10) {
				print "UP k=$k v=$v\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "hsdelete") {
		print("HSDELETE $db.$table op=$op loop=$loop\n");
		$cli->open_index(1, $db, $table, '', '');
		for (my $i = 0; $i < $loop; ++$i) {
			my $k = "" . int(rand($tablesize));
			print "k=$k\n";
			my $r = $cli->execute_delete(1, $op, [ $k ], 1, 0);
			if ($verbose >= 10) {
				print "DEL k=$k\n";
			}
			print "$i/$loop\n" if $i % 1000 == 0;
		}
	} elsif ($action eq "verify") {
		verify_do();
	}
}

sub verify_do {
	my ($fail_cnt, $ok_cnt) = (0, 0);
	my $sth = $dbh->prepare("select v from $db.$table where k = ?");
	use FileHandle;
	my $fh = new FileHandle($file, "r");
	while (my $line = <$fh>) {
		chomp($line);
		my @vec = split(/\t/, $line);
		my $k = $vec[3];
		my $v = $vec[7];
		next if (!defined($k) || !defined($v));
		# print "$k $v\n";
		$sth->execute($k);
		my $aref = $sth->fetchrow_arrayref();
		if (!defined($aref)) {
			print "FAILED: $k notfound\n";
			++$fail_cnt;
		} else {
			my $gv = $aref->[0];
			if ($gv ne $v) {
				print "FAILED: $k got=$gv expected=$v\n";
				++$fail_cnt;
			} else {
				print "OK: $k $v $gv\n" if $verbose >= 10;
				++$ok_cnt;
			}
		}
	}
	print "OK=$ok_cnt FAIL=$fail_cnt\n";
}

sub get_conf {
	my ($key, $def) = @_;
	my $val = $conf{$key};
	if ($val) {
		print "$key=$val\n";
	} else {
		$val = $def;
		$def ||= '';
		print "$key=$def(default)\n";
	}
	return $val;
}

sub get_createtbl_moreflds_str {
	my $s = "";
	for (my $j = 0; $j < $num_moreflds; ++$j) {
		$s .= ",$moreflds_prefix$j varchar(30)";
	}
	return $s;
}

sub get_select_moreflds_str {
	my $s = "";
	for (my $i = 0; $i < $num_moreflds; ++$i) {
		$s .= ",$moreflds_prefix$i";
	}
	return $s;
}

sub get_insert_moreflds_str {
	my $s = "";
	for (my $i = 0; $i < $num_moreflds; ++$i) {
		$s .= ",?";
	}
	return $s;
}

