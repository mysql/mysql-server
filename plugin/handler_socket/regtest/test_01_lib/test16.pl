#!/usr/bin/perl

# vim:sw=2:ai

# test for date/datetime types

BEGIN {
	push @INC, "../common/";
};

use strict;
use warnings;
use bigint;
use hstest;

my $datetime_types = [
	[ 'DATE', '0000-00-00', '2011-01-01', '9999-12-31' ],
	[ 'DATETIME', 0, '2011-01-01 18:30:25' ],
	[ 'TIME', 0, '18:30:25' ],
	[ 'YEAR(4)', 1901, 2011, 2155 ],
	# [ 'TIMESTAMP', 0, 999999999 ],   # DOES NOT WORK YET
];

my $table = 'hstesttbl';
my $dbh;
for my $rec (@$datetime_types) {
  my ($typ, @vals) = @$rec;
  print("TYPE $typ\n");
  test_one($typ, \@vals);
  print("\n");
}

sub test_one {
  my ($typ, $values) = @_;
  $dbh = hstest::init_testdb();
  $dbh->do(
    "create table $table (" .
    "k $typ primary key, " .
    "v1 varchar(512), " .
    "v2 $typ, " .
    "index i1(v1), index i2(v2, v1)) " .
    "engine = myisam default charset = binary");
  my $hs = hstest::get_hs_connection(undef, 9999);
  my $dbname = $hstest::conf{dbname};
  $hs->open_index(1, $dbname, $table, '', 'k,v1,v2');
  $hs->open_index(2, $dbname, $table, 'i1', 'k,v1,v2');
  $hs->open_index(3, $dbname, $table, 'i2', 'k,v1,v2');
  for my $k (@$values) {
    my $kstr = 's' . $k;
    $hs->execute_single(1, '+', [ $k, $kstr, $k ], 0, 0);
  }
  dump_table();
  for my $k (@$values) {
    my $kstr = 's' . $k;
    my ($rk, $rv1, $rv2);
    my $r;
    $r = $hs->execute_single(1, '=', [ $k ], 1, 0);
    shift(@$r);
    ($rk, $rv1, $rv2) = @$r;
    print "PK[$k] $rk $rv1 $rv2\n";
    $r = $hs->execute_single(2, '=', [ $kstr ], 1, 0);
    shift(@$r);
    ($rk, $rv1, $rv2) = @$r;
    print "I1[$kstr] $rk $rv1 $rv2\n";
    $r = $hs->execute_single(3, '=', [ $k, $kstr ], 1, 0);
    shift(@$r);
    ($rk, $rv1, $rv2) = @$r;
    print "I2[$k, $kstr] $rk $rv1 $rv2\n";
    $r = $hs->execute_single(3, '=', [ $k ], 1, 0);
    shift(@$r);
    ($rk, $rv1, $rv2) = @$r;
    print "I2p[$k] $rk $rv1 $rv2\n";
  }
}

sub dump_table {
  print "DUMP_TABLE_BEGIN\n";
  my $aref = $dbh->selectall_arrayref("select k,v1,v2 from $table order by k");
  for my $row (@$aref) {
    my ($k, $v1, $v2) = @$row;
    $v1 = "[null]" if !defined($v1);
    $v2 = "[null]" if !defined($v2);
    print "$k $v1 $v2\n";
    # print "MISMATCH\n" if ($valmap{$k} ne $v);
  }
  print "DUMP_TABLE_END\n";
}

