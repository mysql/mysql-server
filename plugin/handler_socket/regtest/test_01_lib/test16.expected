TYPE DATE
DUMP_TABLE_BEGIN
0000-00-00 s0000-00-00 0000-00-00
2011-01-01 s2011-01-01 2011-01-01
9999-12-31 s9999-12-31 9999-12-31
DUMP_TABLE_END
PK[0000-00-00] 0000-00-00 s0000-00-00 0000-00-00
I1[s0000-00-00] 0000-00-00 s0000-00-00 0000-00-00
I2[0000-00-00, s0000-00-00] 0000-00-00 s0000-00-00 0000-00-00
I2p[0000-00-00] 0000-00-00 s0000-00-00 0000-00-00
PK[2011-01-01] 2011-01-01 s2011-01-01 2011-01-01
I1[s2011-01-01] 2011-01-01 s2011-01-01 2011-01-01
I2[2011-01-01, s2011-01-01] 2011-01-01 s2011-01-01 2011-01-01
I2p[2011-01-01] 2011-01-01 s2011-01-01 2011-01-01
PK[9999-12-31] 9999-12-31 s9999-12-31 9999-12-31
I1[s9999-12-31] 9999-12-31 s9999-12-31 9999-12-31
I2[9999-12-31, s9999-12-31] 9999-12-31 s9999-12-31 9999-12-31
I2p[9999-12-31] 9999-12-31 s9999-12-31 9999-12-31

TYPE DATETIME
DUMP_TABLE_BEGIN
0000-00-00 00:00:00 s0 0000-00-00 00:00:00
2011-01-01 18:30:25 s2011-01-01 18:30:25 2011-01-01 18:30:25
DUMP_TABLE_END
PK[0] 0000-00-00 00:00:00 s0 0000-00-00 00:00:00
I1[s0] 0000-00-00 00:00:00 s0 0000-00-00 00:00:00
I2[0, s0] 0000-00-00 00:00:00 s0 0000-00-00 00:00:00
I2p[0] 0000-00-00 00:00:00 s0 0000-00-00 00:00:00
PK[2011-01-01 18:30:25] 2011-01-01 18:30:25 s2011-01-01 18:30:25 2011-01-01 18:30:25
I1[s2011-01-01 18:30:25] 2011-01-01 18:30:25 s2011-01-01 18:30:25 2011-01-01 18:30:25
I2[2011-01-01 18:30:25, s2011-01-01 18:30:25] 2011-01-01 18:30:25 s2011-01-01 18:30:25 2011-01-01 18:30:25
I2p[2011-01-01 18:30:25] 2011-01-01 18:30:25 s2011-01-01 18:30:25 2011-01-01 18:30:25

TYPE TIME
DUMP_TABLE_BEGIN
00:00:00 s0 00:00:00
18:30:25 s18:30:25 18:30:25
DUMP_TABLE_END
PK[0] 00:00:00 s0 00:00:00
I1[s0] 00:00:00 s0 00:00:00
I2[0, s0] 00:00:00 s0 00:00:00
I2p[0] 00:00:00 s0 00:00:00
PK[18:30:25] 18:30:25 s18:30:25 18:30:25
I1[s18:30:25] 18:30:25 s18:30:25 18:30:25
I2[18:30:25, s18:30:25] 18:30:25 s18:30:25 18:30:25
I2p[18:30:25] 18:30:25 s18:30:25 18:30:25

TYPE YEAR(4)
DUMP_TABLE_BEGIN
1901 s1901 1901
2011 s2011 2011
2155 s2155 2155
DUMP_TABLE_END
PK[1901] 1901 s1901 1901
I1[s1901] 1901 s1901 1901
I2[1901, s1901] 1901 s1901 1901
I2p[1901] 1901 s1901 1901
PK[2011] 2011 s2011 2011
I1[s2011] 2011 s2011 2011
I2[2011, s2011] 2011 s2011 2011
I2p[2011] 2011 s2011 2011
PK[2155] 2155 s2155 2155
I1[s2155] 2155 s2155 2155
I2[2155, s2155] 2155 s2155 2155
I2p[2155] 2155 s2155 2155

